migration/migration.go

```diff
@@ -82,3 +84,49 @@ func revision0000(txn db.Transaction) error {
 	}
 	return it.Close()
 }
+
+// relocateContractStorageRootKeys moves contract root keys to new locations and deletes the previous locations.
+//
+// Before: the key to the root in a contract storage trie was stored at 1+<contractAddress>.
+// After: the key to the root of the contract storage trie is stored at 3+<contractAddress>.
+//
+// This enables us to remove the db.ContractRootKey prefix.
+func relocateContractStorageRootKeys(txn db.Transaction) error {
+	it, err := txn.NewIterator()
+	if err != nil {
+		return err
+	}
+
+	// Iterate over all entries in the old bucket, moving each one to its new location.
+	oldPrefix := db.Unused.Key()
+	for it.Seek(oldPrefix); it.Valid(); it.Next() {
+		// Stop iterating once we're out of the old bucket.
+		if !bytes.Equal(it.Key()[:len(oldPrefix)], oldPrefix) {
+			break
+		}
+
+		oldKey := it.Key()
+
+		// NOTE: contractAddress is a slice of oldKey, which we must not
+		// modify lest we corrupt the iterator.
+		contractAddress, found := bytes.CutPrefix(oldKey, oldPrefix)
+		if !found {
+			// Should not happen. Otherwise the bounds are not working as they should.
+			return db.CloseAndWrapOnError(it.Close, errors.New("iterator not behaving as expected"))
+		}
+
+		value, err := it.Value()
+		if err != nil {
+			return db.CloseAndWrapOnError(it.Close, err)
+		}
+
+		if err := txn.Set(db.ContractStorage.Key(contractAddress), value); err != nil {
+			return db.CloseAndWrapOnError(it.Close, err)
+		}
+		if err := txn.Delete(oldKey); err != nil {
+			return db.CloseAndWrapOnError(it.Close, err)
+		}
```

Comment:
does iterator behave nicely when we are also deleting the keys as we go

Reply:
Nice catch. From the pebble docs:

> // It is also safe to use an iterator concurrently with modifying its
// underlying DB, if that DB permits modification. However, the resultant
// key/value pairs are not guaranteed to be a consistent snapshot of that DB
// at a particular point in time.

I'll just use txn.Get instead of an iterator.